pkg/controller/certificaterequests/ca/ca_test.go

```diff
@@ -391,3 +399,165 @@ func runTest(t *testing.T, test testT) {
 
 	test.builder.CheckAndFinish(err)
 }
+
+func TestCA_Sign(t *testing.T) {
+	rsaPair, err := pki.GenerateRSAPrivateKey(2048)
+	require.NoError(t, err)
+	rsaCSR := generateCSR(t, rsaPair)
+
+	tests := map[string]struct {
+		givenCASecret    *corev1.Secret
+		givenCAIssuer    cmapi.GenericIssuer
+		givenCR          *cmapi.CertificateRequest
+		assertSignedCert func(t *testing.T, got *x509.Certificate)
+		wantErr          string
+	}{
+		"when the CertificateRequest has the duration field set, it should appear as notAfter on the signed ca": {
+			givenCAIssuer: gen.Issuer("issuer-1", gen.SetIssuerCA(cmapi.CAIssuer{
+				SecretName: "secret-1",
+			})),
+			givenCR: gen.CertificateRequest("cr-1",
+				gen.SetCertificateRequestCSR(rsaCSR),
+				gen.SetCertificateRequestIssuer(cmmeta.ObjectReference{
+					Name:  "issuer-1",
+					Group: certmanager.GroupName,
+					Kind:  "Issuer",
+				}),
+				gen.SetCertificateRequestDuration(&metav1.Duration{
+					Duration: 30 * time.Minute,
+				}),
+			),
+			givenCASecret: gen.SecretFrom(gen.Secret("secret-1"), gen.SetSecretNamespace("default"), gen.SetSecretData(secretDataFor(t, rsaPair,
+				&x509.Certificate{
+					SerialNumber: big.NewInt(1234),
+					IsCA:         true,
+				},
+			))),
+			assertSignedCert: func(t *testing.T, got *x509.Certificate) {
+				// The notAfter field uses a precision of 1 second. That's
+				// why we truncate the expected time.
+				expectNotAfter := time.Now().UTC().Truncate(1 * time.Second).Add(30 * time.Minute)
+				assert.Equalf(t, expectNotAfter, got.NotAfter, "time mismatch, expect='%s', got='%s'", expectNotAfter.String(), got.NotAfter.String())
```

Comment:
I suspect you want this:
```diff
-				expectNotAfter := time.Now().UTC().Truncate(1 * time.Second).Add(30 * time.Minute)
-				assert.Equalf(t, expectNotAfter, got.NotAfter, "time mismatch, expect='%s', got='%s'", expectNotAfter.String(), got.NotAfter.String())
+				expectNotAfter := time.Now().UTC().Truncate(1 * time.Second).Add(30 * time.Minute).stripMono()
+				gotNotAfter := got.NotAfter.stripMono()
+				assert.Equalf(t, expectNotAfter, gotNotAfter, "time mismatch, expect='%s', got='%s'", expectNotAfter.String(), gotNotAfter.String())
```
to address: https://github.com/jetstack/cert-manager/pull/3475#issuecomment-789781659

Reply:
Yes, I might want to check that `got.NotAfter` is withing `[expectNotAfter - 2 second, expectNotAfter+ 2 second]` or something like that 😅

Reply:
In the long term, if the API is malleable, maybe you change:
https://github.com/jetstack/cert-manager/blob/923eec2d1bce1e2bc865fb7c5557c38ad0bd045f/pkg/util/pki/csr.go#L320-L321 so that you can provide a time provider instead of being forced to live in real time.

n.b. even if you don't do that, someone should fix it so that `time.Now()` is only called once. It's introducing drift between the two function calls which is arguably wrong.